### PR TITLE
Update Almond to 2.0.0

### PR DESCRIPTION
**Pull request by gcampax**
Also, change the version scheme of the add-on to be the Almond
version number + a monotonically increasing build number, so we
can convert back the add-on version to the Almond version, for
bug triaging.

### CODE REVIEW

File: almond/DOCS.md [hunk 1]
```diff
@@ -27,14 +27,11 @@ Got questions?
...
 In case you've found a bug, please [open an issue on our GitHub][issue].
 
-[discord]: https://discord.gg/c5DvZ4e
-[forum]: https://community.home-assistant.io
-[issue]: https://github.com/home-assistant/hassio-addons/issues
-[reddit]: https://reddit.com/r/homeassistant
-[repository]: https://github.com/hassio-addons/repository
+[discord]: https://discord.gg/anthtR4
+[forum]: https://community.almond.stanford.edu
+[issue]: https://github.com/stanford-oval/almond-server/issues
```
Comment (frenck) [inline almond/DOCS.md:37]: The core add-ons are part of Home Assistant, and thus should be handled here.

Please revert this change. It is OK to link to upstream, but we want to handle and collect issues here. The project carries the responsibility for these add-ons and will redirect users upstream when needed. Additionally, it is very confusing to guide people to different places for different add-ons.

File: almond/Dockerfile [hunk 2]
```diff
@@ -5,50 +5,77 @@ FROM ${BUILD_FROM}
...
     THINGENGINE_HOME="/data/almond-server" \
-    THINGENGINE_HOST_BASED_AUTHENTICATION="local-ip"
+    THINGENGINE_HAS_REVERSE_PROXY=true \
+    THINGENGINE_HOST_BASED_AUTHENTICATION=insecure
 
 WORKDIR /opt/almond
 ARG ALMOND_VERSION
 
+RUN adduser --disabled-password almond && \
+   chown almond:almond /opt/almond
```
Comment (frenck) [inline almond/Dockerfile:15]: We generally don't run under a different user. This might cause issues with things like restoring snapshots, as the permissions on those are not guaranteed. Have you taken care of that? If so, how?

File: almond/build.json [hunk 3]
```diff
@@ -1,12 +1,11 @@
...
   "build_from": {
-    "aarch64": "homeassistant/aarch64-base:3.13",
-    "amd64": "homeassistant/amd64-base:3.13",
-    "i386": "homeassistant/i386-base:3.13",
-    "armv7": "homeassistant/armv7-base:3.13",
-    "armhf": "homeassistant/armhf-base:3.13"
+    "aarch64": "homeassistant/aarch64-base-debian:buster",
+    "amd64": "homeassistant/amd64-base-debian:buster",
+    "armhf": "homeassistant/armhf-base-debian:buster",
+    "armv7": "homeassistant/armv7-base-debian:buster"
```
Comment (frenck) [inline almond/build.json:6]: Why move away from Alpine?

File: almond/config.json [hunk 4]
```diff
@@ -1,17 +1,16 @@
 {
-  "name": "Almond",
-  "version": "1.1.2",
-  "slug": "almond",
-  "description": "The home server version of Almond",
+  "name": "Almond - Edge",
```
Comment (frenck) [inline almond/config.json:2]: This is not the edge, we only want stable add-on in this repo.

File: almond/config.json [hunk 5]
```diff
@@ -1,17 +1,16 @@
...
+  "version": "2.0.0-1",
```
Comment (frenck) [inline almond/config.json:3]: Version numbers of add-ons are decoupled with the software it wraps.

The `-1` addition is not something we use.

```suggestion
  "version": "2.0.0",
```

File: almond/rootfs/etc/services.d/almond/run [hunk 6]
```diff
@@ -7,6 +7,13 @@ export THINGENGINE_BASE_URL
 # Set the Ingress URL as Almond base URL for correct handling
 THINGENGINE_BASE_URL=$(bashio::addon.ingress_entry)
 
+mkdir -p /data/almond-server
+if [ ! -f /data/almond-server/prefs.db ]; then
+    # Skip authentication handling
+    echo '{"server-login":{"password":"x","salt":"x","sqliteKeySalt":"x"}}' \
+      > /data/almond-server/prefs.db
```
Comment (frenck) [inline almond/rootfs/etc/services.d/almond/run:14]: If this file contains fixed content, it can be added as a fixed file. It doesn't need to be "created" from a bash file.

Additionally, this is not runtime/service part, but an init part as it seems. Please put this into an init script.

File: almond/README.md [hunk 7]
```diff
@@ -1,10 +1,15 @@
...
 
-[Almond] For Home Servers.
+## The Open-source, Privacy-Preserving Voice Assistant.
 
 ![Supports aarch64 Architecture][aarch64-shield] ![Supports amd64 Architecture][amd64-shield] ![Supports armhf Architecture][armhf-shield] ![Supports armv7 Architecture][armv7-shield] ![Supports i386 Architecture][i386-shield]
 
-The Open, Privacy-Preserving Virtual Assistant.
+[Almond] provides a complete solution to turn your Home Assistant
+into a smart speaker. It supports playing music, controlling your
+IoTs, checking the weather, and other simple questions.
```
Comment (frenck) [inline almond/README.md:9]: Spelling error: IoT is already multiple.

```suggestion
[Almond] provides a complete solution to turn your Home Assistant
into a smart speaker. It supports playing music, controlling your
IoT, checking the weather, and other simple questions.
```

File: almond/README.md [hunk 8]
```diff
@@ -1,10 +1,15 @@
...
+
+To use Almond, you should connect a microphone and speaker to the
```
Comment (frenck) [inline almond/README.md:11]: Right now, the add-on doesn't require that. Does that change?

File: almond/config.json [hunk 9]
```diff
@@ -1,17 +1,16 @@
...
+  "slug": "edge",
```
Comment (frenck) [inline almond/config.json:4]: We cannot change the slug

File: almond/config.json [hunk 10]
```diff
@@ -1,17 +1,16 @@
...
+  "description": "The open-source, privacy-preserving voice assistant",
   "url": "https://github.com/home-assistant/hassio-addons/blob/master/almond",
-  "arch": ["armv7", "armhf", "aarch64", "amd64", "i386"],
+  "arch": ["armv7", "aarch64", "amd64"],
```
Comment (frenck) [inline almond/config.json:7]: Any specific reason they are dropped?
I don't see any breaking change or information for the end-user about this in the changelog?

File: almond/build.json [hunk 11]
```diff
@@ -1,12 +1,11 @@
 {
   "build_from": {
-    "aarch64": "homeassistant/aarch64-base:3.13",
-    "amd64": "homeassistant/amd64-base:3.13",
-    "i386": "homeassistant/i386-base:3.13",
-    "armv7": "homeassistant/armv7-base:3.13",
-    "armhf": "homeassistant/armhf-base:3.13"
+    "aarch64": "homeassistant/aarch64-base-debian:buster",
+    "amd64": "homeassistant/amd64-base-debian:buster",
+    "armhf": "homeassistant/armhf-base-debian:buster",
```
Comment (frenck) [inline almond/build.json:5]: `armhf` was dropped according to the add-on configuration, but still in here?

File: almond/README.md [hunk 12]
```diff
@@ -1,10 +1,15 @@
 # Home Assistant Add-on: Almond
 
-[Almond] For Home Servers.
+## The Open-source, Privacy-Preserving Voice Assistant.
```
Comment (frenck) [inline almond/README.md:3]: There is no need to make this an heading.

File: almond/Dockerfile [hunk 13]
```diff
@@ -5,50 +5,77 @@ FROM ${BUILD_FROM}
...
+        "https://github.com/stanford-oval/almond-server" . && \
+    sudo -u almond npm ci && \
+    sudo -u almond npm install --no-package-lock --ignore-scripts snowboy && \
+    cd node_modules/snowboy/ && \
+    sudo -u almond PYTHON=python2 npx node-pre-gyp clean configure && \
+    sudo -u almond make -C build/ && \
+    rm -fr /home/almond/.cache && \
+    rm -fr /home/almond/.npm && \
+    mkdir /home/almond/.cache \
+    && apt-get purge -y --auto-remove \
```
Comment (frenck) [inline almond/Dockerfile:69]: Please make the format consistent with the rest of our Dockerfile.

File: almond/Dockerfile [hunk 14]
```diff
@@ -5,50 +5,77 @@ FROM ${BUILD_FROM}
...
-        --run \
-    && yarn global remove modclean \
+    && npm config set unsafe-perm \
     \
-    && yarn cache clean \
-    && apk del --no-cache \
-        .build-dependencies \
+    && sudo -u almond git clone -b "${ALMOND_VERSION}" --depth 1 \
+        "https://github.com/stanford-oval/almond-server" . && \
+    sudo -u almond npm ci && \
```
Comment (frenck) [inline almond/Dockerfile:61]: Sudo?

**Review by frenck — CHANGES_REQUESTED** [review]

File: almond/CHANGELOG.md [hunk 15]
```diff
@@ -1,62 +1,70 @@
...
-- Initial version
+# Changelog
+
+## 2.0.0-1
+
+- Almond was updated to 2.0.0: this is a major release that
+  brings significant changes to the supported skills. Additional
+  details are in the release notes: https://wiki.almond.stanford.edu/en/release-planning/two-point-oh
+- Voice support is now included in this addon, and the Ada add-on
+  is not required (in fact, it should not be enabled at the same time).
```
Comment (frenck) [inline almond/CHANGELOG.md:9]: If that is critical, please ensure that during startup and provide a message to the end-user in the logs.

File: almond/Dockerfile (hunk 2 above)
Comment (gcampax) [reply almond/Dockerfile:15]: It's unsafe to run as root, especially cause you're not using rootless docker. It seems pretty obvious to me.

File: almond/build.json (hunk 3 above)
Comment (gcampax) [reply almond/build.json:6]: Some of the prebuilt software we use depends on glibc.

File: almond/config.json (hunk 5 above)
Comment (gcampax) [reply almond/config.json:3]: It's a mistake to decouple version numbers from upstream software, and one that I hope to avoid with this addon. You and us both need to know what version of Almond is associated with the specific add-on build, otherwise triaging bugs becomes a nightmare.
Even if you consider yourself a "distro" (bad, outdated model) you'll want to track the upstream version and the build number separately.

File: almond/rootfs/etc/services.d/almond/run (hunk 6 above)
Comment (gcampax) [reply almond/rootfs/etc/services.d/almond/run:14]: IIRC there was a race condition during initialization if this was done in the init script. Also, it cannot be a fixed file because it lives in the data partition.

File: almond/README.md (hunk 8 above)
Comment (gcampax) [reply almond/README.md:11]: Yes, the new addon is quite different than the old one, and uses the audio directly.

File: almond/config.json (hunk 10 above)
Comment (gcampax) [reply almond/config.json:7]: They are dropped because they do not build correctly, due to missing packages or binary blobs for those architectures. For example, the version of node was updated to 12, and there are no builds of node 12 for armv6.

File: almond/Dockerfile (hunk 13 above)
Comment (gcampax) [reply almond/Dockerfile:69]: What does that mean?

File: almond/Dockerfile (hunk 14 above)
Comment (gcampax) [reply almond/Dockerfile:61]: Yes. npm is very picky about running as root, and will generally break. So the source is downloaded and compiled as the regular user instead of root. We could use a USER directive but that would split the image into multiple layers and probably make it larger.

File: almond/CHANGELOG.md (hunk 15 above)
Comment (gcampax) [reply almond/CHANGELOG.md:9]: How do I check if a different add-on is enabled?

File: almond/DOCS.md (hunk 1 above)
Comment (frenck) [reply almond/DOCS.md:37]: > We want users to reach out to upstream directly, to report issues and suggestions.

We don't. This was not a request, but a requirement of core integrations. Please adjust as requested.

File: almond/Dockerfile (hunk 2 above)
Comment (frenck) [reply almond/Dockerfile:15]: That didn't answer my questions.

File: almond/build.json (hunk 3 above)
Comment (frenck) [reply almond/build.json]: So why not rebuild it?

File: almond/config.json (hunk 5 above)
Comment (frenck) [reply almond/config.json:3]: > It's a mistake to decouple version numbers from upstream software

It is not. please adjust.

File: almond/rootfs/etc/services.d/almond/run (hunk 6 above)
Comment (frenck) [reply almond/rootfs/etc/services.d/almond/run:14]: Alternatively, store a file outside of the data folder and move it into place.

File: almond/README.md (hunk 8 above)
Comment (frenck) [reply almond/README.md:11]: So it cannot be used without? That would be a big breaking change.

File: almond/config.json (hunk 10 above)
Comment (frenck) [reply almond/config.json:7]: > node 12 for armv6

There is. Also available on Alpine.

File: almond/Dockerfile (hunk 13 above)
Comment (frenck) [reply almond/Dockerfile]: Code styling is completely different. Please adjust it to match the code styling we use in our Dockerfile in our project.

File: almond/Dockerfile (hunk 14 above)
Comment (frenck) [reply almond/Dockerfile]: The linter fails at this as well. There is no reason to use sudo here.

File: almond/CHANGELOG.md (hunk 15 above)
Comment (frenck) [reply almond/CHANGELOG.md:9]: Using the Supervisor API capabilities, this API is documented on our developers website.

File: almond/Dockerfile (hunk 14 above)
Comment (frenck) [reply almond/Dockerfile]: > npm is very picky about running as root, and will generally break.

That is odd, we have quite a lot of add-ons that need NPM that don't show that problem. Any specific reason why this became a problem in this case?

File: almond/build.json (hunk 3 above)
Comment (gcampax) [reply almond/build.json]: I think the issue was snowboy (which uses glibc symbols on arm64) - that's a binary blob, there is nothing we can do about it.

File: almond/Dockerfile (hunk 2 above)
Comment (gcampax) [reply almond/Dockerfile:15]: I don't know if snapshots work with a separate user (although I expect they will if the user ID doesn't change from one image to the next, which it shouldn't).

Now, reading the docker more carefully, the user is not actually used as runtime. It is only used at build time. I didn't write this part of the docker file originally, but I expect it has to do with npm which is broken when run as root. In that case, snapshots should be fine because files in the data file system will be owned by root. Is it ok to leave the non-root user here then?

File: almond/README.md (hunk 8 above)
Comment (gcampax) [reply almond/README.md:11]: It can be used without audio, it's just not very useful. I can clarify this in the readme. Do you have a suggested wording?

File: almond/config.json (hunk 10 above)
Comment (gcampax) [reply almond/config.json:7]: On debian, we need use the nodesource repository for node 12, and that doesn't have node for armv6. I mean, you saw the build failure, right?

File: almond/Dockerfile (hunk 2 above)
Comment (frenck) [reply almond/Dockerfile:15]: If the user isn't used ad runtime, this stuff all isn't really needed and obsolete to begin with.
Please remove it.

File: almond/build.json (hunk 3 above)
Comment (frenck) [reply almond/build.json]: Why are we adding a closed source in here? Are we allowed to distribute it even?
How does this work for the future? Snowboy has been shutdown?

File: almond/Dockerfile (hunk 2 above)
Comment (gcampax) [reply almond/Dockerfile:15]: Again, I think it *is* needed because npm does not run as root, and the build will not run correctly without this. I can test if you insist - but please keep in mind we went through a lot of iterations to make this work, so we have learned what it takes.

File: almond/build.json (hunk 3 above)
Comment (gcampax) [reply almond/build.json]: Yes snowboy is distributable. Snowboy is currently discontinued but it works. There are no open-source wake-word capabilities (Picovoice, used by Ada, is not open-source either, and it has no good general wake word).
In the future, we'll likely change the wakeword and the implementation, but it probably won't be open-source either.

File: almond/Dockerfile (hunk 2 above)
Comment (frenck) [reply almond/Dockerfile:15]: As said, we use it with Z-Wave JS without issue. For my personal add-ons; Z-Wave 2 MQTT, NGinx Proxy Manager, Jupyter Lab (extensions), Node-RED, Home Panel.

So I'm not sure why it shouldn't work here?

File: almond/config.json (hunk 10 above)
Comment (frenck) [reply almond/config.json:7]: Yep, but now learned about Snowboy 👍 
So I guess no choice at this point.

File: almond/Dockerfile (hunk 2 above)
Comment (gcampax) [reply almond/Dockerfile:15]: Is it npm 6? And do those packages require "prepare" steps? In my experience, npm 6 will not run "prepare" steps when it is run as root, which results in a botched build.

Comment (gcampax) [reply almond/Dockerfile:15]: Now, reading through upstream node issues (https://github.com/npm/cli/issues/2062 https://github.com/npm/cli/issues/1287 https://github.com/npm/cli/issues/1214 and others) it seems the bug only occurs with git dependencies. The released version of Almond has no git dependencies, so it's possible that we can remove the non-root user here. Development versions of Almond do have git dependencies though, and I'd rather keep the docker build as close as possible between development and stable versions, so we don't introduce bugs when we build the stable release.

Comment (frenck) [reply almond/Dockerfile:15]: Yes NPM 6.14+ on them all, including native extensions build with gyp

File: almond/README.md (hunk 8 above)
Comment (frenck) [reply almond/README.md:11]:
```suggestion
To use Almond as a voice assistant, you should connect a microphone and speaker to the
```

File: almond/rootfs/etc/services.d/almond/run (hunk 6 above)
Comment (frenck) [reply almond/rootfs/etc/services.d/almond/run:14]: This comment has not been addressed yet.

**Comment by pvizeli** [comment]
Only for clarification, Ada was never a replacement for Almond's voice assist. It was a concept to show how you can build a voice assistant based on tts/stt API of Home Assistant which allows to plug&play different provider of such service.

**Comment by gcampax** [comment]
Yeah I don't mind that Ada continues to exist with that purpose. I know people have custom intents built with Home Assistant's conversation integration and those don't work through Almond. But the common case is for people to combine Ada and Almond (in fact, it is often conflated in social media and support forums) so we should document the change.

**Review by frenck — APPROVED** [review]
Tested it locally, including an upgrade path for existing users, which seems to work fine.

As for audio, I did get output, so that works. Input didn't work on my end, however, I blame my $1.50 USB mini mic for that, I don't see a reason for it being a general issue (@balloob and some others also tested the upstream version, which also showed it working).

So from my end, I think it is OK 👍
